sig-api-machinery/design-docs/admission-control-webhooks.md

```diff
@@ -0,0 +1,934 @@
+*   Ensures dynamic mutations happen before all validations.
+*   Ensures dynamic validations happen after all mutations.
+*   Users don't need to reason about the static ones, just the ones they add.
+
+Validation continues to happen after all the admission controllers (e.g. after mutating webhooks, static admission controllers, and non-mutating admission controllers.)
+
+**TODO**: we should move ResourceQuota after Validation, e.g. as described in #1137.  However, this is a longstanding bug and likely a larger change than can be done in 1.9, and should be revisited along with a larger quota redesign.  Therefore, it is not a requirement for Beta,
+
+
+## Parallel vs Serial 
+
+The main reason for parallel is reducing latency due to round trip and conversion.  We think this can often mitigated by consolidating multiple webhooks shared by the same project into one.
+
+Reasons not to allow parallel are complexity of reasoning about concurrent patches, and CRD not supporting PATCH.
+
+`ValidatingAdmissionWebhook `is already parallel,  and there are no responses to merge.  Therefore, it stays parallel.
+
+`MutatingAdmissionWebhook `will run in serial, to ensure conflicts are resolved deterministically.
+
+The order is the sort order of all the WebhookConfigs, by name, and by index within the Webhooks list.  
+
+We don't plan to make mutating webhooks parallel at this time, but we will revisit the question in the future and decide before going to GA.If a strong need arises for parallel mutating webhooks, we will carefully consider adding a dependency graph.the API will allow adding a partial order later as follows:  Add a "parallelStage" field to WebhookConfig object.  Any objects for which this field is unset all run in serial first.  Then ones which have the phase set are run.  All the ones with the same value run in parallel, and different ones in the order based on the numeric (or lexicographical) order by values).
+
+
+## Good Error Messages	 
+
+When a webhook is persistently failing to allow e.g. pods to be created, then the error message from the apiserver must show which webhook failed.
+
+When a core controller, e.g. ReplicaSet, fails to make a resources, it must send a helpful event that is visible in `kubectl describe` for the controlling resources, saying the reason create failed.
+
+
+## Conversion 
+
+The apiserver will send versioned objects to the webhook to avoid churn in the webhook code. The current registration API has an Rule.[APIVersions](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/admissionregistration/types.go#L86) field, which is used to match against the version in the endpoint url, and only matching requests will be forwarded to the webhook. The current API isn't able to specify what version the APIServer should convert the request body to before forwarding it to the webhook. So we propose two ways to modify the API.
```

Comment:
You modified this section heavily and I agree that it could use some work.  Rather than try to force it through here, we can simply say in this pull that, "The apiserver must send versioned objects to the webhook.  The design of exactly how will be a followup."

Reply:
Yeah and I have a bunch more to say on this. I split it into two sections as there's actually two things to discuss.